### ISSUE TITLE
Occur "Invariant Violation" error when use IE11 mode

### ISSUE THREAD

**Issue by YanYi2017** [pnp/pnpjs#1105]
### Category
- [ ] Enhancement
- [x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [   2.0.3    ]

Please specify what version(s) of SharePoint you are targeting: [  Online   ]

### Expected / Desired Behavior / Question
Running well after using IE11 mode in IE11.

### Observed Behavior
There are errors happened, and the page turn to blank.

### Steps to Reproduce
Build an application customizer extension, and follow [this doc](https://pnp.github.io/pnpjs/concepts/ie11-mode/) to use IE11 mode, and then deploy it to SharePoint. When run it in ie11, it work fine at first, but then an error occur and the page turn into blank.

#### Code
![image](https://user-images.githubusercontent.com/34807478/76302820-9b4c3e80-62fb-11ea-8655-7e74400b2980.png)  

#### package.json
![image](https://user-images.githubusercontent.com/34807478/76303732-2bd74e80-62fd-11ea-81c5-9944d085d26c.png)

#### Error info
![image](https://user-images.githubusercontent.com/34807478/76302898-bc149400-62fb-11ea-8870-2aee5395f907.png)  
&nbsp;
![image](https://user-images.githubusercontent.com/34807478/76300918-786c5b00-62f8-11ea-9143-f9cbd39567b1.png)  
&nbsp;
![image](https://user-images.githubusercontent.com/34807478/76301000-8f12b200-62f8-11ea-94ba-28c298328f66.png)  

#### Detailed error info in ie console log
[Console log.txt](https://github.com/pnp/pnpjs/files/4311683/Console.log.txt)

**Comment by koltyakov** [comment]
Not sure if it actually comes from PnPjs.
React validation detects unhandled promise which was rejected.
Could you try Debug mode with the non-minified build? It could potentially expose the real source of the issue.

**Comment by YanYi2017** [comment]
@koltyakov Tanks for your reply. Do you mean follow [this way](https://pnp.github.io/pnpjs/debugging/) to debug? And gulp without "--ship"  to get the non-minified build?

**Comment by koltyakov** [comment]
I was thinking of this debug technique https://docs.microsoft.com/en-us/sharepoint/dev/spfx/debug-modern-pages

**Comment by YanYi2017** [comment]
Ok, I will try it later. Thanks

**Comment by siddharth-vaghasia** [comment]
@YanYi2017 ....this was happening to me as well for IE 11 in one of the project.... Issue has nothing to do with pnp js....For me it was an import statement(refer below) that was not giving compile error but generating invariant Violation at runtime.....  check all your import statements in your ts and tsx file...if there is something like below this error will be generated.

![image](https://user-images.githubusercontent.com/9557557/76675522-4dc42000-65e0-11ea-8549-249430481080.png)

hope this helps..!!

**Comment by AkshayKishorChauhan** [comment]
I have faced such issue in IE after using  Pnp 2.0 with IE Mode Enabled,I have deployed my solution on 3 pages
1.Classic Page - it works fine on classic page
2. Modern Page on  Communication Site - It works fine if you turn off comments of Modern page and then deploy the solution.
3.Modern Page on Team Site - not working 
Whats happening is The solution loads then everything goes blank.I was trying to deploy default solution that comes @microsoft/sharepoint generator with Pnp 2.0 and IE11 Mode enabled.Please help if anybody knows why this is happening.

**Comment by patrick-rodgers** [comment]
@YanYi2017 or @AkshayKishorChauhan  if you can provide a repo that shows the error we can clone we can try and take a look, but this feels like something external to the library. Or if you can pinpoint what is causing the error we can have a look.

**Comment by AkshayKishorChauhan** [comment]
@patrick-rodgers Thank you so much for looking into this,I have uploaded the code to https://github.com/AkshayKishorChauhan/helloGithub
This solution is also giving error in IE on Modern Communication page.Please let me know in case you need more info.

**Comment by AkshayKishorChauhan** [comment]
Hi @patrick-rodgers ,Thank for the reply I have imported the polyfills https://github.com/AkshayKishorChauhan/helloGithub/blob/master/src/webparts/helloGithub/components/HelloGithub.tsx at line 5.
I am not sure where else to add.

**Comment by patrick-rodgers** [comment]
I would try moving it to the main webpart class. I have never put that import in a component before. And I am far from a React expert but is it common to do async operations within componentDid* calls?

I have in the past done all my data calls in the upper most webpart and updated components via the model. Not what you are doing is wrong, just an approach I haven't taken.

**Comment by AkshayKishorChauhan** [comment]
I have tried moving polyfills to main webpart class,still the same issue.We can call operations within componentDidMount.I have just made an sample solution you can modify it as needed.The only solution I have found is to revert back to Pnp 1.x. And this issue is reproduced by many of my friends who are using pnp.

**Comment by harshcse1025** [comment]
Same issue I'm facing which @AkshayKishorChauhan is facing once we deploy our web part on app catalog and when we add the webpart in modern Pages of SharePoint. It get showed and then disappears.

**Comment by kunj-sangani** [comment]
Hi All,

I was able to replicate the issue and I Found that 
**import "core-js/stable/symbol";**
in below mentioned file is causing the issue by removing this it worked for me 
https://github.com/pnp/pnpjs/blob/version-2/tools/polyfill-ie11/index.ts

Below are the modules which needs to be installed
"core-js": "3.4.7",
"es6-map": "0.1.5",
"whatwg-fetch": "3.0.0"

Below are the imports which needs to be used
import "core-js/stable/array/from";
import "core-js/stable/array/fill";
import "core-js/stable/array/iterator";
import "core-js/stable/promise";
import "core-js/stable/reflect";
import "es6-map/implement";
import "whatwg-fetch";


I will raise a PR for the change in the @pnp/polyfill-ie11

**Comment by patrick-rodgers** [comment]
Well, doesn't that remove the polyfill for symbol which was added to resolve #996. How deeply have you tested your changes? This is not an error that we see often and we have been unable to reproduce. Until we can I am reluctant to merge your PR which removes a fix for a previous bug.

**Comment by kunj-sangani** [comment]
For reproducing the issue you can use a sample project which I have created 
[pnpjs2ie11Test](https://github.com/kunj-sangani/pnpjs2ie11Test)

Deploy it Modern Team Site where we are able to reproduce the issue
Hope this helps

**Comment by dave-flex** [comment]
I have also came across this issue, we've been developing several extensions and they all fail in IE11 compatibility mode when we go into edit mode, everything else works flawlessly, right now I'm going back to pnp/sp@1.3.11 but I still have the hope to find the specific issue, I'm also using React for my extensions and what I have found is polyfill v2.0.1 fails with react, it gives this error when going to react error-decoder "Objects are not valid as a React child (found: object with keys {$$typeof, type, key, ref, props, _owner})." going down to v1.0.2 seems to be the solution but i'm still not quite certain on this, I'll keep you posted or be pending on this matter for any updates.

**Comment by patrick-rodgers** [comment]
@dave-flex - if you can please report the issue you describe with details that would be good. Thanks!

**Comment by harshcse1025** [comment]
Hi Patrik any update on @kunj-sangani and my issue. kunj have already shared the solution as well

**Comment by juliemturner** [comment]
So I cloned your repo and tested it and had no issues with the webpart.  Loaded fine and other than normally seen console errors with IE11 nothing was reported.

![image](https://user-images.githubusercontent.com/7570936/79278553-e537bf80-7e79-11ea-94c4-52e77a72394b.png)

**Comment by dave-flex** [comment]
my @pnp/sp, odata, common and logging version: 2.0.4
my @pnp/polyfill-ie11 version: 2.0.1
the extension loads in the page correctly but when i click on the edit button in the page after a few seconds this is what I get:
![image](https://user-images.githubusercontent.com/62066601/79278723-f2e94700-7e71-11ea-83a2-60b2e2568578.png)

**Comment by juliemturner** [comment]
Ok, but that's an issue with react.js, not PnPJS. We were simply trying to validate that we can/cannot reproduce an issue with the PnPJS polyfill. As I understand it there are polyfills for react.js to make it work on ie11 but to be honest I have not worked with them.

**Comment by kunj-sangani** [comment]
> So I cloned your repo and tested it and had no issues with the webpart. Loaded fine and other than normally seen console errors with IE11 nothing was reported.
> 
> ![image](https://user-images.githubusercontent.com/7570936/79278553-e537bf80-7e79-11ea-94c4-52e77a72394b.png)

@juliemturner  To replicate the issue we have to use the webpart in SharePoint Teams Site and not in SharePoint Communication site

@dave-flex  If you can try removing the comment section present in the page this should help 
remove the comment section from chrome and then test in IE

**Comment by patrick-rodgers** [comment]
Did you visit t[he link](https://reactjs.org/docs/error-decoder.html/?invariant=31&args%5B%5D=object%20with%20keys%20%7B%24%24typeof%2C%20type%2C%20key%2C%20ref%2C%20props%2C%20_owner%7D&args%5B%5D=) and search on the problem?

As Julie said things appear to be working for us.

**Comment by dave-flex** [comment]
Hi, I probably had to add some more detail on what i was facing so I opened [this](https://github.com/pnp/pnpjs/issues/1161) to provide further information on the matter. 

Thanks!

**Comment by patrick-rodgers** [comment]
After much thought we determined we don't have a great way to support every possible combination of polyfills folks might need. To that end we have an [updated statement around our polyfills](https://pnp.github.io/pnpjs/concepts/polyfill/#general-statement-on-polyfills). We have included the ability to selectively import just the polyfills you need based on your application. There are just too many different libraries and scenarios for a single package to universally work without conflicts. Thank you to all for the discussion, we appreciate your ideas. If you have continued thoughts please share them in a new issue. Thanks!
